File: crates/treeman-core/src/envfile.rs
//! Read shell-style env files (`.env`, `.env.testing`, etc.).
//!
//! Mirrors the subset of the dotenv spec used by Laravel, Symfony, Node
//! (`dotenv`), Python (`python-dotenv`), Ruby (`dotenv`), and sqlx-cli.
//! Supports:
//!   * `KEY=VALUE` and `export KEY=VALUE`
//!   * Single- or double-quoted values
//!   * Inline `# comment` after unquoted values
//!   * `${OTHER}` variable expansion against the same file's prior keys
//!     and the process env
//!
//! Does NOT support:
//!   * Multiline values (Laravel doesn't emit them; if needed later, add
//!     a heredoc detector)
//!   * Command substitution `$(cmd)` (security hazard)

use std::collections::HashMap;
use std::path::Path;

use anyhow::{Context, Result};

#[derive(Debug, Clone, Default)]
pub struct EnvFile {
    pub vars: HashMap<String, String>,
    pub source: Option<std::path::PathBuf>,
}

impl EnvFile {
    pub fn get(&self, key: &str) -> Option<&str> {
        self.vars.get(key).map(|s| s.as_str())
    }
    pub fn merge(&mut self, other: EnvFile) {
        for (k, v) in other.vars {
            self.vars.entry(k).or_insert(v);
        }
    }
}

pub fn parse(input: &str) -> EnvFile {
    let mut out = EnvFile::default();
    for line in input.lines() {
        let line = line.trim_start();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let line = line.strip_prefix("export ").unwrap_or(line);
        let Some(eq) = line.find('=') else {
            continue;
        };
        let key = line[..eq].trim().to_string();
        let mut chars = key.chars();
        let first_ok = chars
            .next()
            .map(|c| c.is_ascii_alphabetic() || c == '_')
            .unwrap_or(false);
        if !first_ok || !chars.all(|c| c.is_ascii_alphanumeric() || c == '_') {
            continue;
        }
        let raw = line[eq + 1..].trim();
        let value = strip_quotes_and_comment(raw);
        let expanded = expand(&value, &out.vars);
        out.vars.insert(key, expanded);
    }
    out
}

pub fn read(path: &Path) -> Result<EnvFile> {
    let txt = std::fs::read_to_string(path)
        .with_context(|| format!("read env file {}", path.display()))?;
    let mut env = parse(&txt);
    env.source = Some(path.to_path_buf());
    Ok(env)
}

/// Load `path` if it exists; otherwise return an empty EnvFile.
pub fn read_optional(path: &Path) -> EnvFile {
    if path.is_file() {
        read(path).unwrap_or_default()
    } else {
        EnvFile::default()
    }
}

/// Walk through the conventional env files for a repo, with later files
/// overriding earlier (matches Laravel's `.env` → `.env.testing` order).
pub fn read_layered<P: AsRef<Path>>(paths: &[P]) -> EnvFile {
    let mut out = EnvFile::default();
    for p in paths {
        let one = read_optional(p.as_ref());
        for (k, v) in one.vars {
            out.vars.insert(k, v);
        }
    }
    out
}

fn strip_quotes_and_comment(s: &str) -> String {
    let s = s.trim();
    if (s.starts_with('"') && s.ends_with('"') && s.len() >= 2)
        || (s.starts_with('\'') && s.ends_with('\'') && s.len() >= 2)
    {
        return s[1..s.len() - 1].to_string();
    }
    // Strip inline comments on unquoted values: `VAL # note`
    if let Some(hash) = s.find(" #") {
        return s[..hash].trim_end().to_string();
    }
    s.to_string()
}

fn expand(s: &str, locals: &HashMap<String, String>) -> String {
    // Cheap `${VAR}` interpolation, no recursion.
    let mut out = String::with_capacity(s.len());
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'$' && i + 1 < bytes.len() && bytes[i + 1] == b'{' {
            if let Some(end) = s[i + 2..].find('}') {
                let key = &s[i + 2..i + 2 + end];
                if let Some(v) = locals.get(key).cloned().or_else(|| std::env::var(key).ok()) {
                    out.push_str(&v);
                } else {
                    out.push_str(&s[i..i + 2 + end + 1]);
                }
                i += 2 + end + 1;
                continue;
            }
        }
        out.push(bytes[i] as char);
        i += 1;
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_kv() {
        let e = parse("FOO=bar\nBAZ=qux\n");
        assert_eq!(e.get("FOO"), Some("bar"));
        assert_eq!(e.get("BAZ"), Some("qux"));
    }

    #[test]
    fn quoted_values() {
        let e = parse(
            r#"A="hello world"
B='single quoted'
"#,
        );
        assert_eq!(e.get("A"), Some("hello world"));
        assert_eq!(e.get("B"), Some("single quoted"));
    }

    #[test]
    fn export_prefix() {
        let e = parse("export FOO=bar\n");
        assert_eq!(e.get("FOO"), Some("bar"));
    }

    #[test]
    fn inline_comment_unquoted_only() {
        let e = parse("FOO=bar # trailing\nBAZ=\"x # not a comment\"\n");
        assert_eq!(e.get("FOO"), Some("bar"));
        assert_eq!(e.get("BAZ"), Some("x # not a comment"));
    }

    #[test]
    fn comments_skipped() {
        let e = parse("# top\nFOO=bar\n#mid\nBAZ=qux\n");
        assert_eq!(e.vars.len(), 2);
    }

    #[test]
    fn expand_local() {
        let e = parse("FOO=hello\nBAR=${FOO}/world\n");
        assert_eq!(e.get("BAR"), Some("hello/world"));
    }

    #[test]
    fn ignores_invalid_keys() {
        let e = parse("12 = no\nFOO=ok\n  =empty\n");
        assert_eq!(e.vars.len(), 1);
        assert_eq!(e.get("FOO"), Some("ok"));
    }
}

File: internal/envfile/envfile.go
// Package envfile reads shell-style env files (`.env`,
// `.env.testing`, etc.). Mirrors the subset of the dotenv spec used
// by Laravel, Symfony, Node, Python, Ruby, sqlx-cli.
//
// Supports KEY=VALUE / `export KEY=VALUE`, single/double quoted
// values, inline `# comment` on unquoted values, and `${OTHER}`
// expansion against earlier keys + the process env. Does NOT support
// multi-line values or command substitution.
package envfile

import (
	"bufio"
	"maps"
	"os"
	"strings"
)

// EnvFile is the parsed key/value map plus the on-disk source for
// provenance.
type EnvFile struct {
	Vars   map[string]string
	Source string // absolute path of the file this was parsed from; "" for inline parse
}

// Get returns the value for key, or empty + false if absent.
func (e EnvFile) Get(key string) (string, bool) {
	v, ok := e.Vars[key]
	return v, ok
}

// Merge layers `other` on top, but only inserts keys not already
// present. Used by ReadLayered when iterating .env → .env.testing
// → etc., since later files override earlier ones (caller handles
// ordering by calling Merge in the reverse order, or by direct
// insertion).
func (e *EnvFile) Merge(other EnvFile) {
	for k, v := range other.Vars {
		if _, exists := e.Vars[k]; !exists {
			e.Vars[k] = v
		}
	}
}

// Parse a string into an EnvFile (no Source).
func Parse(input string) EnvFile {
	out := EnvFile{Vars: map[string]string{}}
	sc := bufio.NewScanner(strings.NewReader(input))
	sc.Buffer(make([]byte, 0, 64*1024), 16*1024*1024)
	for sc.Scan() {
		line := strings.TrimLeft(sc.Text(), " \t")
		if line == "" || line[0] == '#' {
			continue
		}
		line = strings.TrimPrefix(line, "export ")
		before, after, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		key := strings.TrimSpace(before)
		if !validKey(key) {
			continue
		}
		raw := strings.TrimSpace(after)
		value := stripQuotesAndComment(raw)
		expanded := expand(value, out.Vars)
		out.Vars[key] = expanded
	}
	return out
}

// Read parses a file from disk. Returns empty + nil error when the
// file doesn't exist (callers iterate over conventional paths and
// most of them are absent).
func Read(path string) (EnvFile, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return EnvFile{Vars: map[string]string{}}, nil
		}
		return EnvFile{Vars: map[string]string{}}, err
	}
	e := Parse(string(b))
	e.Source = path
	return e, nil
}

// ReadOptional is the no-error variant of Read — non-existent or
// unreadable files yield an empty EnvFile. Use when iterating
// conventional dotenv paths where most are expected to be missing.
func ReadOptional(path string) EnvFile {
	e, _ := Read(path)
	return e
}

// ReadLayered walks the conventional env files for a repo, later
// files overriding earlier (Laravel's `.env` → `.env.testing`
// order). The returned EnvFile's Source field tracks the last file
// that actually existed — used by `resolve` for provenance labels.
func ReadLayered(paths []string) EnvFile {
	out := EnvFile{Vars: map[string]string{}}
	for _, p := range paths {
		one := ReadOptional(p)
		if one.Source != "" {
			out.Source = one.Source
		}
		maps.Copy(out.Vars, one.Vars)
	}
	return out
}

func validKey(key string) bool {
	if key == "" {
		return false
	}
	for i := range len(key) {
		c := key[i]
		isAlpha := (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_'
		isDigit := c >= '0' && c <= '9'
		// First char must be a letter or underscore; digits are allowed
		// only from the second char on.
		if isAlpha || (i > 0 && isDigit) {
			continue
		}
		return false
	}
	return true
}

func stripQuotesAndComment(s string) string {
	s = strings.TrimSpace(s)
	if len(s) >= 2 {
		first, last := s[0], s[len(s)-1]
		if (first == '"' && last == '"') || (first == '\'' && last == '\'') {
			return s[1 : len(s)-1]
		}
	}
	// Strip inline comments on unquoted values: `VAL # note`.
	if before, _, ok := strings.Cut(s, " #"); ok {
		return strings.TrimRight(before, " \t")
	}
	return s
}

// expand resolves `${VAR}` against the file's own prior keys, then
// the process env. No recursion.
func expand(s string, locals map[string]string) string {
	var b strings.Builder
	b.Grow(len(s))
	i := 0
	for i < len(s) {
		if s[i] == '$' && i+1 < len(s) && s[i+1] == '{' {
			end := strings.IndexByte(s[i+2:], '}')
			if end >= 0 {
				key := s[i+2 : i+2+end]
				if v, ok := locals[key]; ok {
					b.WriteString(v)
				} else if v := os.Getenv(key); v != "" {
					b.WriteString(v)
				} else {
					b.WriteString(s[i : i+2+end+1])
				}
				i += 2 + end + 1
				continue
			}
		}
		b.WriteByte(s[i])
		i++
	}
	return b.String()
}
